doc/proposals/upstream-osdk-features-into-controller-runtime.md

```diff
@@ -0,0 +1,150 @@
+
+Ideally the flagset for setting all the logger configurations could also live upstream but given that the controller-runtime’s [pkg/log/zap][cr-pkg-log-zap] allows instantiating multiple zap loggers with different configs, it may not be suitable to have a global flagset that provides a singular configuration for all instantiated loggers.
+This point needs more discussion and it’s possible that the configuration flags may have to live downstream in the SDK.
+
+
+#### Story 4 - Operator Developers can use SDK’s method of building images that run as non-root users in Kubebuilder projects
+
+Kubebuilder scaffolded projects would previously run the operator base images with the user as root. See Kubebuilder’s [pkg/scaffold/v2/dockerfile.go][kb-scaffold-dockerfile].
+
+Operator SDK scaffolds a project Dockerfile such that it runs as non-root by default and allows the operator image to run with arbitrary UIDs on openshift. For details, see the [openshift container guidelines][openshift-container-guidelines] for non-root images, and how the SDK includes the user setup in the image build and uses a custom entrypoint:
+
+- [internal/pkg/scaffold/build_dockerfile.go][sdk-scaffold-dockerfile]
+- [internal/pkg/scaffold/entrypoint.go][sdk-scaffold-entrypoint]
+- [internal/pkg/scaffold/usersetup.go][sdk-scaffold-user-setup]
+
+Kubebuilder should support scaffolding projects that will allow the base image to run as non-root and support arbitrary user ids. 
+Currently with [kubebuilder #983][kubebuilder-983], a non-root base image should be supported by Kubebuilder.
+
+
+#### Story 5 - Operator Developers can use the prometheus-operator’s ServiceMonitor API to configure prometheus to scrape their operator metrics in Kubebuilder projects
+
+The Operator SDK’s [pkg/metrics][sdk-pkg-metrics] has helpers that let’s operators configure and create [ServiceMonitors][service-monitor-doc] that lets a prometheus instance on a cluster target the operator’s Service object that exposes operator metrics.
+
+Instead of having this functionality live in controller-runtime as helpers that can be called to setup ServiceMonitors, this can added as manifests that are scaffolded by Kubebuilder for an operator project. This would be similar to other resources that need to be created alongside the operator and can be customized in the manifest. Upstream issue at [kubebuilder #887][kubebuilder-887].
+
+
+#### Story 6 - Operator Developers have documentation that demonstrates how to create and expose custom operator metrics
+
+Once Kubebuilder supports scaffolding ServiceMonitor manifests, there needs to be a section documented in the [Kubebuilder book][kubebuilder-book] that goes over an example of how an operator should record and expose custom metrics.
```

Comment:
Should we link to the existing documentation in the kubebuilder book about creating custom metrics? I think this is a pretty good start on what we envisioned, though we'll probably need to update it to include info about using ServiceMonitors.

https://book.kubebuilder.io/reference/metrics.html

Reply:
Updated to point to the existing section. And the story now is to extend that documentation once Kubebuilder supports ServiceMonitors.